### PR TITLE
Triggering rasterization of Scroller Content after zoom operation on RS5+

### LINKED ISSUE #136: Multiple unreliable Scroller tests
ScrollerTests.BasicOffsetChanges and ScrollerTests.BasicZoomFactorChanges are unreliable in test passes and have been disabled.

**ScrollerTests.BasicOffsetChanges**
https://microsoft.visualstudio.com/WinUI/_Build?_a=summary&buildId=13513372&tab=ms.vss-test-web.test-result-details
[Error]: System.Exception: Timeout expiration in WaitForEvent. at Windows.UI.Xaml.Tests.MUXControls.ApiTests.ScrollerTests.WaitForEvent(String logComment, EventWaitHandle eventWaitHandle) at Windows.UI.Xaml.Tests.MUXControls.ApiTests.ScrollerTests.ChangeOffsets(Scroller scroller, Double horizontalOffset, Double verticalOffset, ScrollerViewKind viewKind, ScrollerViewChangeKind viewChangeKind, ScrollerViewChangeSnapPointRespect snapPointRespect, Boolean hookViewChanged, Nullable`1 expectedFinalHorizontalOffset, Nullable`1 expectedFinalVerticalOffset) at Windows.UI.Xaml.Tests.MUXControls.ApiTests.ScrollerTests.BasicOffsetChanges()

**ScrollerTests.BasicZoomFactorChanges**
https://microsoft.visualstudio.com/WinUI/_Build?_a=summary&buildId=13513372&tab=ms.vss-test-web.test-result-details
[Error]: System.Exception: Timeout expiration in WaitForEvent. at Windows.UI.Xaml.Tests.MUXControls.ApiTests.ScrollerTests.WaitForEvent(String logComment, EventWaitHandle eventWaitHandle) at Windows.UI.Xaml.Tests.MUXControls.ApiTests.ScrollerTests.ChangeZoomFactor(Scroller scroller, Single zoomFactor, Single centerPointX, Single centerPointY, ScrollerViewKind viewKind, ScrollerViewChangeKind viewChangeKind, Boolean hookViewChanged) at Windows.UI.Xaml.Tests.MUXControls.ApiTests.ScrollerTests.BasicZoomFactorChanges()

### LINKED ISSUE #541: ScrollViewer / Scroller doesn't re-rasterize the content after a zoom
**Describe the bug**

ScrollViewer/Scroller doesn't re-rasterize the content after a zoom

**Steps to reproduce the bug**

1. Use the Microsoft.UI.Xaml.Controls.ScrollViewer control to display some zoomable content that contains text (e.g. TextBlocks)
2. Set ZoomMode="Enabled"
3. Zoom in/out
4. The text isn't re-rendered after the zoom and remains fuzzy

**Expected behavior**

The content may appear fuzzy while a zoom is in progress, but it should re-rasterize when the zoom completes.

**Version Info**

NuGet package version: 

2.1.190319001-prerelease

| Windows 10 version                  | Saw the problem? |
| :--------------------------------- | :-------------------- |
| October 2018 Update (17763)   | Yes   |

| Device form factor | Saw the problem? |
| :-------------------- | :------------------- |
| Desktop                 | Yes   |

**Additional context**

### PR DESCRIPTION
**Pull request by RBrid**
Fixes #541 - Scroller.Content rasterization after zoom operation.
Fixes #136 - Scroller test failure.

The Scroller.Content was not re-rasterizing after a zoom operation, causing the rendering to be blurry. The solution here is to stop the zoom factor expression animation targeting the UIElement.Scale façade property until a batch update was sent to Composition, then re-start the animation. This causes a rasterization.

I added a Scroller test hook to detect animation starts and stops in general, and those zoom factor animation interruptions in particular. All the view change API tests are now at the lookout for those private test hook notifications.

Also fixed:
- a typo in ScrollerOperation ScrollerTests.StartScrollBy, and re-enabled two Scroller tests.
- a bug in Scroller::ValuesChanged so that a non-animated view change is declared completed as soon as the ValuesChanged event is raised.

### CODE REVIEW

File: dev/Scroller/Scroller.cpp [hunk 1]
```diff
@@ -3038,66 +3047,185 @@ void Scroller::SetupTransformExpressionAnimations(
...
+    bool forZoomFactorAnimationInterruption)
 {
     if (content)
     {
         if (SharedHelpers::IsTranslationFacadeAvailable(content))
         {
-            content.StopAnimation(m_translationExpressionAnimation);
+            if (!forZoomFactorAnimationInterruption)
+            {
+                wstring_view scrollPropertyName = GetVisualTargetedPropertyName(ScrollerDimension::Scroll);
```
Comment (StephenLPeters) [inline dev/Scroller/Scroller.cpp]: I'd inline this, but if you don't want to I'd use auto

File: dev/Scroller/Scroller.cpp [hunk 2]
```diff
@@ -3038,66 +3047,185 @@ void Scroller::SetupTransformExpressionAnimations(
...
+
+                content.StopAnimation(m_translationExpressionAnimation);
+                RaiseExpressionAnimationStatusChanged(false /*isExpressionAnimationStarted*/, scrollPropertyName /*propertyName*/);
+            }
+
+            wstring_view zoomFactorPropertyName = GetVisualTargetedPropertyName(ScrollerDimension::ZoomFactor);
```
Comment (StephenLPeters) [inline dev/Scroller/Scroller.cpp]: same

File: dev/Scroller/Scroller.cpp [hunk 3]
```diff
@@ -3038,66 +3047,185 @@ void Scroller::SetupTransformExpressionAnimations(
...
         }
-        else
+        else if (!forZoomFactorAnimationInterruption) // The zoom factor animation interruption is only effective with facades.
         {
             const winrt::Visual contentVisual = winrt::ElementCompositionPreview::GetElementVisual(content);
             if (IsVisualTranslationPropertyAvailable())
             {
-                contentVisual.StopAnimation(GetVisualTargetedPropertyName(ScrollerDimension::Scroll));
-                contentVisual.StopAnimation(GetVisualTargetedPropertyName(ScrollerDimension::ZoomFactor));
+                wstring_view scrollPropertyName = GetVisualTargetedPropertyName(ScrollerDimension::Scroll);
```
Comment (StephenLPeters) [inline dev/Scroller/Scroller.cpp]: >wstring_view [](start = 16, length = 12)

auto const

File: dev/Scroller/Scroller.h [hunk 4]
```diff
@@ -75,6 +75,11 @@ class Scroller :
     static constexpr int s_zoomFactorChangeMinMs{ 50 };
     static constexpr int s_zoomFactorChangeMaxMs{ 1000 };
 
+    // Number of ticks ellapsed before restarting the Scale animation to allow
+    // the Content rasterization to be triggered after the Idle State is reached
+    // or a zoom factor change operation completed.
+    static constexpr int s_zoomFactorAnimationRestartTicks = 4;
```
Comment (StephenLPeters) [inline dev/Scroller/Scroller.h]: What is the logic behind the choice of 4?

Comment (RBrid) [reply dev/Scroller/Scroller.h]: I talked with Keith M. and unfortunately there is no deterministic way for the Scroller to know when the composition layer saw the stoppage to trigger the rasterization. Based on my observations, 4 ticks was always enough.

**Comment by adrientetar** [comment]
@jevansaks Do you plan to put out a new prerelease? I would like to test the improvements in this PR.

**Comment by jevansaks** [comment]
> @jevansaks Do you plan to put out a new prerelease? I would like to test the improvements in this PR.

Thanks for asking! Yes we're planning one for this week, we were just waiting for a couple of in-progress ScrollViewer and TabView features.

**Pull request merged**